README
                            GImage View README
                                                                 for ver 0.2.21
                                                Last Modified : 2003/05/22(Thu)
                                                      Created : 2001/09/08(Sat)
-------------------------------------------------------------------------------
  ** Contents **
    0. What's this?
    1. Introduction
    2. Requirement
    3. Download
    4. Install
    5. Usage
    6. ChangeLog of README
    7. License
    8. Remarks
-------------------------------------------------------------------------------

--------------------
0. What's this?

GImageView is an image viewer which solves a hopless situation under
displaying a lot of images.
It works on X Window System, frame buffer of linux, and Cygwin.



--------------------
1. Introduction

GImageView has following features:

 * Simple, but usefull/flexible user interface:
   Thumbnail window has 3 paned view (directory tree view, thumbnail view and
   image preview), and its layout is fully castomizable by Drag and Drop.
   Images can be also displayed by separated image window.

 * Managing image files:
   It can copy, move and link image files by Drag and Drop.

 * Low waste memory.

 * Also usefull for command line usage:
   If you specify image files and direcotries from comamnd line, image files are
   opened by image window, and directories are opened by thumbnail window
   automatically.

 * Tabbed browsing:
   It can open two or more directories at same time by tabbed thumbnail window
   which looks like tabbed web browser. If you drop files to tab, you can copy,
   move and link files to the directory.

 * Supported following graphich formats:

   JPEG
   PNG
   PNM
   GIF
   TIFF
   XBM
   XPM
   BMP
   PCX
   TGA
   XCF
   MAG
   SVG (librsvg is required)
   MNG (libmng is required)
   WMF (libwmf is required)
   ANI (Gtk+-2 is required)
   RAS (Gtk+-2 is required)

 * It can play animation such as animation GIF and MNG.

 * Supports movie files and sound files using Xine and MPlayer.
   Supported file types are depended on Xine and MPlayer:

   * AVI (MS-MPEG4, DivX, MotionJPEG, Cinepack and so on...)
   * MPEG1, MPEG2, MPEG4, MPEG1 audio layer3
   * Quick Time
   * Real Media
   * Any other unknown types...

 * It can extract compressed file automatically.
   (File name extension is not registed by default. Please open common ->
    filter page of preference dialog and regist it manually).

    gzip, bzip2

 * Supports archive files using external command:

   tar, zip, lha, rar

 * Supports various thumbnail cache types

   Nautilus, Konqueror, GQview, Electric Eyes (Picview), xvpics
   (Thumbnail Managing Standard which supported by latest Nautilus and Konqueror
    will be implemented at next versoin of GImageView)

 * Supports slideshow

 * Threre are 10 types of thumbnail view, and supports vatious sort types.

 * Finding duplicated images

   It can find duplicated image files by file siez, md5sum and similarity,
   and display it on tree view.



--------------------
2. Requirements

You need following environment/libraries to build GImaqeView:

 * ANSI-C, POSIX
 * GLib-1.2 and GTK+-1.2 or later  (http://www.gtk.org/).
   Also support GtK+-2.0 or later.
 * Imlib-1.9.x (ftp://ftp.gnome.org/pub/GNOME/stable/sources/imlib/)
   or
   gdk-pixbuf (ftp://ftp.gnome.org/pub/GNOME/unstable/sources/gdk-pixbuf/)
 * zlib and libpng

Following libraries and programs are optional:

 * GNU file
 * libjpeg
 * libmng-1.0.0 or later
 * librsvg-1.0.0 or later
 * libwmf-0.2.8 or later
 * bzip2-1.0.0 or later
 * xine-lib-0.9.8 or later (1.0-beta10 or later is recomanded)
 * MPlayer-0.90-rc2 or later (0.90 or later is remomanded)
 * GNU tar
 * zip
 * lha
 * rar

GImageView is confirmed on followin environment:

 * RedHat Linux 7.3, 8, 9
 * Mandrake Linux
 * Debian GNU/Linux potate, woody, sarge
 * Gentoo Linux 1.4-rc2
 * Plamo Linux 2.0 (Alpha3 Base), 3.0
 * Kondara MNU/Linux 1.2 and 2.0
 * Momonga Linux
 * Vine Linux (i386 & ppc) 2.1.5, 2.5. 2.6
 * FreeBSD 4.2-RELEASE
 * NetBSD
 * AIX



--------------------
3. Download

You can get it from follwing place include old versions:

  * http://sourceforge.net/project/showfiles.php?group_id=39083

You can get also latest develop version from CVS:

$ export CVSROOT=:pserver:anonymous@cvs.gtkmmviewer.sourceforge.net:/cvsroot/gtkmmviewer
$ cvs login
  (Passward is emty, simply press Enter)
$ cvs -z3 co gimageview

You can update it by following command:

$ cvs -z3 update

Develop version doesn't include configure script. Please execute "autogen.sh"
script which placed top directory of source tree. You need autoconf, automake
and libtool to make it.

In addition to it, please call your attention that CVS version is "latest"
"develop" version. It may includes various bugs and sometimes it may losts
your important data. Please use it at your own risk. If you found bugs, please
report it to us.


--------------------
4. Install

You can build it by following commands on all most all environment.

 $ tar xvfz gimageview-x.y.z.tar.gz
 $ ./configure
 $ make
 $ su
  (Enter passward)
 # make install

Please replace the version number suitably.

If you want to compile it against Imlib, please use "--disable-gdk-pixbuf"
configure option:

 $ ./configure --disable-gdk-pixbuf

If you want to compile against Gtk+2, please use "--with-gtk2"

 $ ./configure --with-gtk2

To compile against linux-fb, please add "gtk+-linux-fb-2.0" argument to
"--with-gtk2" option.

 $ ./configure --with-gtk2="gtk+-linux-fb-2.0"

If you want to enable movie support, please add "--with-xine" or
"--enable-mplayer" options. These options can be used at same time.

 $ ./configure --with-xine --enable-mplayer

In addition to it, if you don't need debug information, you can strip it by
following command.

 # make install-strip


Development of GImageView is still in early stage, so format of preference file
may be changed. Please backup your own ~/.gimv directory when update the
GImageView.



--------------------
5. Usage

You can show one or more image files by following command:

 $ gimv [Image files...]

If you specify one or more directories with a "-d" option, you can show
thumbnails of all images under the directory by thumbnail window.

 $ gimv -d [Directory name]

If you specify both images and directories with a "-d" option, images are
opened by image window, and thumbnails under directories are opened by thumbnail
window.

 $ gimv -d foo01.jpg ~/ ~/images images/foo02.jpg\
      /usr/share/pixmaps/ bar.png ......

If you add a "-t" option, all images are opened by thumbnail window.

Followings are other options:

 Usage: gimv [OPTION...] [Image files or directories...]
   -d, --directory          Scan directories at start up
   -R, --recursive          Scan directories recursively (use with "-d")
   -D, --scan-dot           Read dotfile when scan directory (use with "-d")
   -e, --ignore-ext         Ignore file name extention
   -s, --scale=SCALE        Specify image scale on image window(%)
   -b, --buffer=ON/OFF      Keeo original image on memory or not
   -M, --menubar            Show menu bar on image window
   -T, --toolbar            Show tool bar on image window
   -w, --thumbwin           Open thumbnail window at start up
   -i, --imageview          Open all images by image window
   -t, --thumbview          Open all images by thumbnail window
   -S, --slideshow          Open images by slide show
   -v, --version            Print version infomation
   -h, --help               Show this message

If you want to open a directory from GUI, please double click it on directory
view, or specify it from file open dialog.

If you want to show compressed image files, please open common -> filter page of
preference dialog and regist file name extention like "gz" or "bz2".



--------------------
6. ChangeLog of README

2003/06/01(Thu) Takuro Ashie

 * Fixed typo (English Version)

2003/05/22(Thu) Takuro Ashie

 * Tranlated to English
 * Follow latest version (0.2.21).
 * Remove ChangeLog and TODO

2001/11/06(Tue) nyan2

 * for version 0.1.3
 * add CVS section in "get gimageview" chapter

2001/11/04(Sun) Takuro Ashie

 * Modified TODO
 * Added ChangeLog and minor fix.

2001/11/04(Sun) nyan2

 * for version 0.1.2

2001/10/18(Thu) nyan2

 * Minor change

2001/10/03(Wed) nyan2

 * for version 0.1.1

2001/09/12(Wed) - 2001/09/28(Fri) nyan2

 * for nightly
 * Follow modified features and bugfixes.

2001/09/10(Mon) nyan2

 * for 2001/10/10 nightly
 * Initial import

2001/09/08(Sat) nyan2

 * for 2001/09/08 nightly
 * Stated editing.



--------------------
7. License

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


--------------------
8. Remarks

This document is modified by Nayn2 based on
http://www.homa.ne.jp/~ashie/gimageview/.

Copyright (C) 2001-2003
  Original : Takuro Ashie
  Modified : Nyan2 <t-nyan2@nifty.com>
-------------------------------------------------------------------------------
                                                                  [End Of File]


NEWS
Version-0.3.0

	* test

Version-0.2.27

	* Added some options for slideshow.
	  (Thanks! > Christian Hammers <lathspell@users.sourceforge.net>)
	* Added Xine image loader for movie thumbnail. (experimental)
	* The bug that sometime same files are appeared numbers of times on
	  same thumbnail view has been fixed.
	  (Thanks reporting > shlomif)
	* Enable to open archived files which has "[" in its filename.
	* Other minor bug fixes are made.

Version-0.2.26

	* Improved performance of redrawing thumbnail view.
	* Improved preference of "fit image size to window" feature.
	  (Thanks! > Brent Baccala <baccala@users.sourceforge.net>)
	* Fixed bugs on 64bit envirponment.
	  (Thanks! > Tomoya TAKA <tomoya@olive.plala.or.jp>)
	* Enabled to open EXIF view from image window.
	* Added image scale preferences for slideshow.
	* Removed obsolete summary modes of thumbnail view.
	  ("List" and "Thumbnail" mode)
	* Fixed compile time error on Gtk+-2.4
	* Other minor fixes and improvements.

Version-0.2.25

	* Adapt to latest MPlayer.
	* Xienerama support.
	* "270" orientation label is changed to "-90".
	* Some minor usabilty fix of comment view.

Version-0.2.24

	* "Auto fit without zoom in" feature has been fixed.
	* Added "Fit width" and "Fit height" feature.
	* The bug that sometimes mplayer fails to start playing has been fixed.
	* Fixed double freeing bug of JPEG loader.
	  (Thanks! > sheepman <sheepman@tcn.zaq.ne.jp>)
	* Fixed a bug in DnD support
	  (Thanks! > MINAMI Hirokazu <h_minami@users.sourceforge.net>)
	* Remove trailing slash when create directory.
	  (Thanks! > Douglas Brebner <douglas+mail@fang.demon.co.uk>).
	* Updated libexif to 0.5.11.
	* Other code improvements.

Version-0.2.23

	* Fixed clash bug of slideshow.
	* Plugged memory leak of JPEG loader.
	* Other minor fixes.

Version-0.2.22

	* Added automatic rotation by EXIF data.
	 (contributed from Frank Fischer <frank_fischer@gmx.de>. Thanks!)
	* Fixed detecting file name extension as case insensitive.
	 (Thanks! > Ricky Faulstich <Ricky@Faulstich.com>)
	* Fixed typo (Thanks! > Ricky Faulstich <Ricky@Faulstich.com>)

Version-0.2.21

	* Fixed compile error.

Version-0.2.20

	* Image loaders, I/O streamers, Image View and Thumbnail View Embders
          now are implemented as plugins.
	* Now GImageView can be compiled against linux-fb backend of Gtk+-2.0 by
	  "--with-gtk2=gtk+-linux-fb-2.0" configure opotion.
	* Merged Cygwin patch provided by Masahiro Sakai. Now GImageView can be
	  compiled on Cygwin.
	* Added some key binding to directory view on thumbnail window.
		CTRL + Cursor Left
		CTRL + Cursor Up: Change top directory to parent
		CTRL + Cursor Right
		CTRL + Cursor Down: Change top directory to selected
	* "Thumbnail 2" mode has been renamed to "Album" mode.
	* Added a key binding to "Album" mode.
		SHIFT + Cursor: select the fucesed image and open it.
	* Modified directory list preference to make it easy.
	* "Load thumbnails recursively in one tab" can be attached to mouse
	  button.
	* Now gimv doesn't zoom in by default when fit to frame automatically.
	* JPEG loader is added. Most of its codes are taken from libgnome-ui.
	  Creating JPEG thumbnails are much faster than GdkPixbuf.
	* Added "Slideshow (selected)" and "Slideshow (from the selected)".
	* Now you can override MPlayer options by
	  $(datadir)/gimageview/mplayerrc.
	* Enabled to play audio with visualizer
	  (Visualizer feature is supported by Xine view only).
	* Set focus to the suitable thumbnail when files are deleted.
	* Added rotation indicator in toolbar of image window.
	* Two new duplicates finders are added. One is "md5sum" and another one
	  is "File Size".
	* Clash bug of Xine view has been fixed.
	* Other minor fixes.


src/help.c
/* -*- Mode: C; tab-width: 3; indent-tabs-mode: nil; c-basic-offset: 3 -*- */

/*
 * GImageView
 * Copyright (C) 2001 Takuro Ashie
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 * $Id$
 */

#include "gimageview.h"

#include <stdlib.h>
#include <string.h>
#include <sys/utsname.h>
#include <gdk-pixbuf/gdk-pixbuf-features.h>

#include "gimv_icon_stock.h"
#include "gimv_plugin.h"
#include "gimv_text_win.h"
#include "prefs.h"
#include "help.h"
#include "utils_char_code.h"
#include "utils_file.h"
#include "utils_file_gtk.h"
#include "utils_menu.h"

#define DOC_HTML_DIR "html"
#define DOC_TEXT_DIR "text"
#define GIMV_MANUAL_FILE "index.html"

typedef struct GimvInfoWin_Tag
{
   GtkWidget *window;
   GtkWidget *scrolled_win;
   GtkWidget *text_box;
} GimvInfoWin;


/* callback functions */
static void cb_open_manual             (gpointer   data,
                                        guint      action,
                                        GtkWidget *widget);
static void cb_open_text               (GtkWidget *menuitem,
                                        gchar     *filename);
static void cb_gimv_info_win_ok_button (GtkWidget *widget,
                                        GtkWidget *window);
static void cb_open_info               (gpointer   data,
                                        guint      action,
                                        GtkWidget *widget);

/* other private functions */
static gchar       *get_doc_dir_name     (const gchar *lang,
                                          const gchar *type);
static GtkWidget   *get_dirlist_sub_menu (const gchar *dir,
                                          gpointer     func,
                                          GList      **filelist);


static GimvInfoWin info_win;

static gchar *license = 
N_("Copyright (C) 2001 %s <%s>\n\n"
   "This program is free software; you can redistribute it and/or modify\n"
   "it under the terms of the GNU General Public License as published by\n"
   "the Free Software Foundation; either version 2, or (at your option)\n"
   "any later version.\n\n"

   "This program is distributed in the hope that it will be useful,\n"
   "but WITHOUT ANY WARRANTY; without even the implied warranty of\n"
   "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"
   "See the GNU General Public License for more details.\n\n"

   "You should have received a copy of the GNU General Public License\n"
   "along with this program; if not, write to the Free Software\n"
   "Foundation, Inc., 59 Temple Place - Suite 330, Boston,\n"
   "MA 02111-1307, USA.");

static gchar *authors = 
N_("Main Program :\n"
   "    Takuro Ashie <ashie@homa.ne.jp>\n"
   "Document :\n"
   "    Nyan2 <t-nyan2@nifty.com>\n"
   "Logo :\n"
   "    eins <eins@milk.freemail.ne.jp>\n"
   "Translate :\n"
   "\n"
   "Special Thanks :\n"
   "    horam\n"
   "    TAM\n"
   "    Hiroyuki Komatsu\n"
   "    Kazuki Iwamoto\n"
   "    katsu\n"
   "    kourin\n"
   "    jissama\n"
   "    shitamori\n"
   "    knee\n"
   "    matsu\n"
   "    Shlomi Fish\n"
   "    Jin Suh\n"
   "    sheepman\n"
   "    MINAMI Hirokazu\n"
   "    Brent Baccala\n"
   "    Christian Hammers\n"
   "    And all GImageView users");

static gchar *system_info = NULL;

static gchar *plugin_info = NULL;

GtkItemFactoryEntry gimvhelp_menu_items[] =
{
   {N_("/_Manual"),               NULL, cb_open_manual, 0, NULL},
   {N_("/_Document"),             NULL, NULL,           0, "<Branch>"},
   {N_("/_Document/_HTML"),       NULL, NULL,           0, "<Branch>"},
   {N_("/_Document/Plain _Text"), NULL, NULL,           0, "<Branch>"},
   {N_("/_About"),                NULL, cb_open_info,   0, NULL},
   {NULL, NULL, NULL, 0, NULL},
};


GList *html_filelist = NULL;
GList *text_filelist = NULL;


/******************************************************************************
 *
 *   Callback functions for menubar.
 *
 ******************************************************************************/
static void
cb_open_manual(gpointer data, guint action, GtkWidget *widget)
{
   gchar *dir, manual[MAX_PATH_LEN], *cmd, buf[BUF_SIZE];
   dir = get_doc_dir_name (NULL, DOC_HTML_DIR);
   g_snprintf (manual, MAX_PATH_LEN, "%s/%s", dir, GIMV_MANUAL_FILE);
   if (file_exists(manual)) {
      g_snprintf (buf, BUF_SIZE, conf.web_browser, manual);
      cmd = g_strconcat (buf, " &", NULL);
      system (cmd);
      g_free (cmd);
   }
   g_free (dir);
}


static void
cb_open_text (GtkWidget *menuitem, gchar *filename)
{
   gchar *cmd = NULL;

   g_return_if_fail (filename);

   if (!conf.text_viewer_use_internal && conf.text_viewer) {
      cmd = g_strconcat (conf.text_viewer, " ", filename, " &", NULL);
      system (cmd);
   } else {
      GtkWidget *text_win = gimv_text_win_new ();
      gimv_text_win_load_file (GIMV_TEXT_WIN (text_win), filename);
      gtk_widget_show (text_win);
   }

   g_free (cmd);
}


static void
cb_open_html (GtkWidget *menuitem, gchar *filename)
{
   gchar buf[BUF_SIZE], *cmd;

   g_return_if_fail (filename);

   if (conf.web_browser && *conf.web_browser) {
      g_snprintf (buf, BUF_SIZE, conf.web_browser, filename);
      cmd = g_strconcat (buf, " &", NULL);
      system (cmd);
      g_free (cmd);
   }
}


static void
cb_progurl_clicked (GtkWidget *widget, gchar *url)
{
   gchar buf[BUF_SIZE], *cmd;

   g_return_if_fail (url);

   if (conf.web_browser && *conf.web_browser) {
      g_snprintf (buf, BUF_SIZE, conf.web_browser, url);
      cmd = g_strconcat (buf, " &", NULL);
      system (cmd);
      g_free (cmd);
   }
}


static void
set_copyleft_str (void)
{
   GtkTextBuffer *buffer;
   gchar buf[BUF_SIZE];

   g_snprintf (buf, BUF_SIZE, _(license),
               GIMV_PROG_AUTHOR, GIMV_PROG_ADDRESS);

   buffer = gtk_text_view_get_buffer (GTK_TEXT_VIEW (info_win.text_box));
   gtk_text_buffer_set_text (buffer, "\0", -1);

   if (*buf)
      gtk_text_buffer_set_text (buffer, buf, -1);
}


static void
cb_gimv_info_change_text (GtkWidget *widget, gchar *text)
{
   GtkTextBuffer *buffer;

   g_return_if_fail (info_win.text_box);

   if (!text) {
      set_copyleft_str ();
      return;
   }

   buffer = gtk_text_view_get_buffer (GTK_TEXT_VIEW (info_win.text_box));
   gtk_text_buffer_set_text (buffer, "\0", -1);

   if (text && *text)
         gtk_text_buffer_set_text (buffer, text, -1);
}


static void
cb_gimv_info_win_ok_button (GtkWidget *widget, GtkWidget *window)
{
   gtk_widget_destroy (window);
}


static void
cb_open_info (gpointer data, guint action, GtkWidget *widget)
{
   gimvhelp_open_info_window ();
}


/******************************************************************************
 *
 *   Private functions.
 *
 ******************************************************************************/
static gchar *
get_doc_dir_name (const gchar *lang, const gchar *type)
{
   gchar *dir, *lang_fallback, *tmp;

   g_return_val_if_fail (type && *type, NULL);

   if (!lang)
      lang = get_lang ();

   dir = g_strconcat (DOCDIR, "/", type, "/", lang, NULL);

   if (isdir (dir)) return dir;

   lang_fallback = g_strdup (lang);
   tmp = strchr (lang_fallback, '.');
   if (tmp) {
      *tmp = '\0';
      g_free (dir);
      dir = g_strconcat (DOCDIR, "/", type, "/", lang_fallback, NULL);
      g_free (lang_fallback);
      if (isdir (dir)) {
         return dir;
      }
   } else {
      g_free (lang_fallback);
   }

   if (strlen (lang) > 2 && lang[2] == '_') {
      lang_fallback = g_strdup (lang);
      lang_fallback[2] = '\0';
      g_free (dir);
      dir = g_strconcat (DOCDIR, "/", type, "/", lang_fallback, NULL);
      g_free (lang_fallback);
      if (isdir (dir)) {
         return dir;
      }
   }

   return NULL;
}


static GtkWidget *
get_dirlist_sub_menu (const gchar *dir, gpointer func, GList **filelist)
{
   GtkWidget *menu = NULL, *menuitem;
   GList *node;

   g_return_val_if_fail (filelist, NULL);

   if (!dir) return NULL;;

   menu = gtk_menu_new();

   if (!*filelist)
      get_dir (dir, GETDIR_FOLLOW_SYMLINK, filelist, NULL);

   node = *filelist;
   while (node) {
      gchar *filename = node->data;

      node = g_list_next (node);

      if (!filename) continue;

      gchar *basename = g_path_get_basename (filename);
      menuitem = gtk_menu_item_new_with_label (basename);
      g_free (basename);
      g_signal_connect (G_OBJECT (menuitem), "activate",
                        G_CALLBACK (func), filename);
      gtk_menu_append (GTK_MENU (menu), menuitem);
      gtk_widget_show (menuitem);

   }

   return menu;
}


static void
help_plugin_info_append (GList *list)
{
   GList *node;
   gchar buf[BUF_SIZE];

   /* g_return_if_fail (list); */

   if (!plugin_info)
      plugin_info = g_strdup ("");

   if (!list) return;

   node = list;
   while (node) {
      GModule *module = node->data;
      gchar *tmpstr;
      g_snprintf (buf, BUF_SIZE,
                  _("Plugin Name : %s\n"
                    "Version : %s\n"
                    "Author : %s\n\n"),
                  gimv_plugin_get_name (module),
                  gimv_plugin_get_version_string (module),
                  gimv_plugin_get_author (module));
      tmpstr = plugin_info;
      plugin_info = g_strconcat (plugin_info, buf, NULL);
      g_free (tmpstr);

      node = g_list_next (node);
   }
}


/******************************************************************************
 *
 *   Public functions.
 *
 ******************************************************************************/
GtkWidget *
gimvhelp_create_menu (GtkWidget *window)
{
   GtkWidget *menu = NULL, *menuitem = NULL, *html_submenu, *text_submenu;
   GtkItemFactory *ifactory;
   guint n_menu_items;
   gchar *dir, manual[MAX_PATH_LEN];

   n_menu_items = sizeof (gimvhelp_menu_items)
      / sizeof (gimvhelp_menu_items[0]) - 1;
   menu = menu_create_items(window, gimvhelp_menu_items,
                            n_menu_items, "<HelpSubMenu>", NULL);

   dir = get_doc_dir_name (NULL, DOC_HTML_DIR);
   html_submenu = get_dirlist_sub_menu (dir, (gpointer) cb_open_html,
                                        &html_filelist);
   if (html_submenu)
      menu_set_submenu (menu, "/Document/HTML", html_submenu);
   g_free (dir);
   dir = NULL;

   dir = get_doc_dir_name (NULL, DOC_TEXT_DIR);
   text_submenu = get_dirlist_sub_menu (dir, (gpointer) cb_open_text,
                                        &text_filelist);
   if (text_submenu)
      menu_set_submenu (menu, "/Document/Plain Text", text_submenu);
   g_free (dir);
   dir = NULL;

   dir = get_doc_dir_name (NULL, DOC_HTML_DIR);
   g_snprintf (manual, MAX_PATH_LEN, "%s/%s", dir, GIMV_MANUAL_FILE);
   g_free (dir);
   if (!file_exists(manual)) {
      ifactory = gtk_item_factory_from_widget (menu);
      menuitem  = gtk_item_factory_get_item (ifactory, "/Manual");
      gtk_widget_set_sensitive (menuitem, FALSE);
   }

   return menu;
}


GtkWidget *
gimvhelp_create_info_widget (void)
{
   GtkWidget *vbox, *hbox1, *hbox2, *tmpvbox;
   GtkWidget *label;
   GtkWidget *button, *radio;
   GtkWidget *scrolledwin, *text;
   GtkWidget *frame, *frame_vbox;
   GtkWidget *logo;
   gchar buf[BUF_SIZE] /* , alt_string[BUF_SIZE] */;
   struct utsname utsbuf;

   /* create info string */
   if (!system_info) {
      uname(&utsbuf);
      g_snprintf (buf, BUF_SIZE,
                  _("Operating System : %s %s %s\n"
                    "GTK+ version : %d.%d.%d\n"
                    /* "libpng version : %s\n" */),
                  utsbuf.sysname, utsbuf.release, utsbuf.machine,
                  gtk_major_version, gtk_minor_version, gtk_micro_version
                  /*, png_get_header_ver (NULL)*/);
#if 0
#ifdef ENABLE_MNG
      g_snprintf (alt_string, sizeof (alt_string) / sizeof (gchar),
                  _("libmng version : %s\n"),
                  mng_version_text ());
      strncat (buf, alt_string, BUF_SIZE - strlen (buf));
#endif
      g_snprintf (alt_string, sizeof (alt_string) / sizeof (gchar),
                  _("gdk-pixbuf version : %s\n"),
                  gdk_pixbuf_version);
      strncat (buf, alt_string, BUF_SIZE - strlen (buf));
#ifdef ENABLE_SVG
      g_snprintf (alt_string, sizeof (alt_string) / sizeof (gchar),
                  _("librsvg version : %s\n"),
                  librsvg_version);
      strncat (buf, alt_string, BUF_SIZE - strlen (buf));
#endif
#ifdef ENABLE_XINE
      /*
      g_snprintf (alt_string, sizeof (alt_string) / sizeof (gchar),
                  _("Xine version : %s\n"),
                  xine_get_str_version());
      strncat (buf, alt_string, BUF_SIZE - strlen (buf));
      */
#endif
#endif
      system_info = g_strdup (buf);
   }

   if (!plugin_info) {
#if 0
      gint idx;
      const gchar *type;

      for (idx = 0; (type = gimv_plugin_type_get (idx)) != NULL; idx++) {
         GList *list = NULL;

         list = gimv_plugin_get_list (type);
         help_plugin_info_append (list);
      }
#else
      {
         GList *list = NULL;

         list = gimv_plugin_get_list (NULL);
         help_plugin_info_append (list);
      }
#endif
   }

   /* create content widget */
   vbox = gtk_vbox_new (FALSE, 0);

   frame = gtk_frame_new (NULL);
   gtk_container_set_border_width(GTK_CONTAINER(frame), 0);
   gtk_box_pack_start(GTK_BOX(vbox), frame, FALSE, FALSE, 0);
   gtk_widget_show (frame);

   frame_vbox = gtk_vbox_new (FALSE, 0);
   gtk_container_set_border_width (GTK_CONTAINER(frame), 5);
   gtk_container_add (GTK_CONTAINER (frame), frame_vbox);
   gtk_widget_show (frame_vbox);

   logo = gimv_icon_stock_get_widget ("gimageview");
   gtk_box_pack_start (GTK_BOX (frame_vbox), 
                       logo, FALSE, TRUE, 5);
   gimv_icon_stock_free_icon ("gimageview");
   gtk_widget_show (logo);

   tmpvbox = frame_vbox;

   /* Program name & Copyright */ 
   label = gtk_label_new (_(GIMV_PROG_VERSION));
   gtk_box_pack_start (GTK_BOX (tmpvbox), 
                       label, FALSE, FALSE, 0);
   gtk_widget_show (label);

   /* Web Site Button */
   hbox1 = gtk_hbox_new (TRUE, 0);
   gtk_box_pack_start (GTK_BOX (tmpvbox), 
                       hbox1, FALSE, FALSE, 0);
   gtk_widget_show (hbox1);
   hbox2 = gtk_hbox_new (FALSE, 0);
   gtk_box_pack_start (GTK_BOX (hbox1), 
                       hbox2, TRUE, FALSE, 0);
   gtk_widget_show (hbox2);

   label = gtk_label_new (_("Web Site: "));
   gtk_box_pack_start (GTK_BOX (hbox2), 
                       label, FALSE, FALSE, 0);
   gtk_widget_show (label);

   button = gtk_button_new ();
   gtk_button_set_relief (GTK_BUTTON (button), GTK_RELIEF_NONE);
   label = gtk_label_new (GIMV_PROG_URI);
   gtk_container_add (GTK_CONTAINER (button), label);
   gtk_widget_show (label);
   g_signal_connect (G_OBJECT (button), "clicked",
                     G_CALLBACK (cb_progurl_clicked),
                     GIMV_PROG_URI);
   gtk_box_pack_start (GTK_BOX (hbox2), 
                       button, FALSE, FALSE, 0);
   gtk_widget_show (button);

   /* Infomation Text Box */
   scrolledwin = gtk_scrolled_window_new (NULL, NULL);
   info_win.scrolled_win = scrolledwin;
   gtk_scrolled_window_set_policy (GTK_SCROLLED_WINDOW(scrolledwin),
                                   GTK_POLICY_AUTOMATIC, GTK_POLICY_AUTOMATIC);
   gtk_scrolled_window_set_shadow_type(GTK_SCROLLED_WINDOW(scrolledwin),
                                       GTK_SHADOW_IN);
   gtk_box_pack_start (GTK_BOX (vbox),
                       scrolledwin, TRUE, TRUE, 0);
   gtk_container_set_border_width (GTK_CONTAINER (scrolledwin), 5);
   gtk_widget_show (scrolledwin);

   text = gtk_text_view_new ();
   gtk_container_add (GTK_CONTAINER (scrolledwin), text);
   info_win.text_box = text;
   set_copyleft_str ();
   gtk_widget_show (text);

   /* Radio Button */
   hbox1 = gtk_hbox_new (FALSE, 0);
   gtk_box_pack_start (GTK_BOX (vbox), 
                       hbox1, FALSE, FALSE, 0);
   gtk_widget_show (hbox1);
   hbox2 = gtk_hbox_new (FALSE, 0);
   gtk_box_pack_start (GTK_BOX (hbox1), 
                       hbox2, TRUE, FALSE, 0);
   gtk_widget_show (hbox2);

   radio = gtk_radio_button_new_with_label (NULL, _("License"));
   g_signal_connect (G_OBJECT (radio), "clicked",
                     G_CALLBACK (cb_gimv_info_change_text), NULL);
   gtk_box_pack_start (GTK_BOX (hbox2), radio, FALSE, FALSE, 0);
   gtk_widget_show (radio);

   radio = gtk_radio_button_new_with_label_from_widget (GTK_RADIO_BUTTON (radio),
                                                        _("Authors"));
   g_signal_connect (G_OBJECT (radio), "clicked",
                     G_CALLBACK (cb_gimv_info_change_text),
                     _(authors));
   gtk_box_pack_start (GTK_BOX (hbox2), radio, FALSE, FALSE, 0);
   gtk_widget_show (radio);

   radio = gtk_radio_button_new_with_label_from_widget (GTK_RADIO_BUTTON (radio),
                                                        _("System Info"));
   g_signal_connect (G_OBJECT (radio), "clicked",
                     G_CALLBACK (cb_gimv_info_change_text), system_info);
   gtk_box_pack_start (GTK_BOX (hbox2), radio, FALSE, FALSE, 0);
   gtk_widget_show (radio);

   radio = gtk_radio_button_new_with_label_from_widget (GTK_RADIO_BUTTON (radio),
                                                        _("Plugin Info"));
   g_signal_connect (G_OBJECT (radio), "clicked",
                     G_CALLBACK (cb_gimv_info_change_text), plugin_info);
   gtk_box_pack_start (GTK_BOX (hbox2), radio, FALSE, FALSE, 0);
   gtk_widget_show (radio);

   return vbox;
}


void
gimvhelp_open_info_window (void)
{
   GtkWidget *widget, *window, *button;
   gchar buf[BUF_SIZE];

   /* window */
   window = gtk_dialog_new ();
   info_win.window = window;
   gtk_container_set_border_width (GTK_CONTAINER (GTK_DIALOG (window)->vbox), 0);
   gtk_window_set_position (GTK_WINDOW (window), GTK_WIN_POS_CENTER);
   g_snprintf (buf, BUF_SIZE, _("About %s"), GIMV_PROG_NAME);
   gtk_window_set_wmclass(GTK_WINDOW(window), "about", "GImageView");
   gtk_window_set_title (GTK_WINDOW (window), buf); 
   gtk_window_set_default_size (GTK_WINDOW (window), 500, 400);

   /* main content */
   widget = gimvhelp_create_info_widget ();
   gtk_box_pack_start (GTK_BOX (GTK_DIALOG (window)->vbox), 
                       widget, TRUE, TRUE, 0);
   gtk_widget_show (widget);

   /* OK Button */
   button = gtk_button_new_with_label (_("OK"));
   gtk_box_pack_start (GTK_BOX (GTK_DIALOG (window)->action_area), 
                       button, TRUE, TRUE, 0);
   g_signal_connect (G_OBJECT (button), "clicked",
                     G_CALLBACK (cb_gimv_info_win_ok_button),
                     window);
   GTK_WIDGET_SET_FLAGS(button,GTK_CAN_DEFAULT);
   gtk_widget_show (button);

   gtk_widget_show (window);
   gimv_icon_stock_set_window_icon (window->window, "gimv_icon");

   gtk_widget_grab_focus (button);

   gtk_grab_add (window);
}
